wire-moshi-support/src/main/java/com/squareup/wire/WireJsonAdapterFactory.java

```diff
@@ -0,0 +1,50 @@
+package com.squareup.wire;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.Set;
+import javax.annotation.Nullable;
+import okio.ByteString;
+
+/**
+ * A {@link JsonAdapter.Factory} that allows Wire messages to be serialized and deserialized using
+ * the Moshi Json library.
+ *
+ * <pre>
+ * Moshi moshi = new Moshi.Builder()
+ *   .add(new WireJsonAdapterFactory())
+ *   .build();
+ * </pre>
+ *
+ * The resulting {@link Moshi} instance will be able to serialize and deserialize any Wire {@link
+ * Message} type, including extensions and unknown field values. The JSON encoding is intended to be
+ * compatible with the
+ * <a href="https://code.google.com/p/protobuf-java-format/">protobuf-java-format</a>
+ * library. Note that version 1.2 of that API has a
+ * <a href="https://code.google.com/p/protobuf-java-format/issues/detail?id=47">bug</a>
+ * in the way it serializes unknown fields, so we use our own approach for this case.
+ */
+public final class WireJsonAdapterFactory implements JsonAdapter.Factory {
+  @Nullable
+  @Override
+  public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+    if (!annotations.isEmpty()) {
+      return null;
+    }
+
+    if (Types.getRawType(type).equals(ByteString.class)) {
+      return new ByteStringJsonAdapter();
+    }
+    if (Message.class.isAssignableFrom(Types.getRawType(type))) {
+      return new MessageJsonAdapter(moshi, type);
+    }
+    if (Types.getRawType(type).equals(Object.class)) {
+      JsonAdapter<Object> delegate = moshi.nextAdapter(this, Object.class, annotations);
+      return new SegregatedNumbersAdapter(delegate);
```

Comment:
I wonder if there’s a way we can do this only on recursive calls required by the MessageJsonAdapter

Reply:
not a problem, anymore